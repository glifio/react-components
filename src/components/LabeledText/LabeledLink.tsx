import styled from 'styled-components'
import PropTypes from 'prop-types'

import { LabeledText } from '.'
import { SmartLink } from '../SmartLink'
import { CopyText } from '../CopyText'
import { IconNewTab } from '../Icons'
import { Colors } from '../theme'

const LabeledLinkEl = styled.div`
  display: flex;
  gap: 0.25em;
  line-height: 1.5;

  > a {
    color: ${props => props.color};
    text-decoration: none;

    &:hover {
      color: ${props => props.color};
      text-decoration: underline;
    }

    svg {
      transition: 0.24s ease-in-out;
      vertical-align: middle;

      &:hover {
        transform: scale(1.25);
        fill: ${Colors.PURPLE_MEDIUM};
      }
    }
  }
`

export const LabeledLink = ({
  label,
  color,
  href,
  linkText,
  copyText,
  disableLink,
  hideCopy,
  hideCopyText,
  useNewTabIcon
}: LabeledLinkProps) => {
  return (
    <LabeledText label={label}>
      <LabeledLinkEl color={color}>
        {disableLink || useNewTabIcon ? (
          <span>{linkText}</span>
        ) : (
          <SmartLink href={href}>{linkText}</SmartLink>
        )}
        {useNewTabIcon && (
          <SmartLink href={href}>
            <IconNewTab />
          </SmartLink>
        )}
        {!hideCopy && copyText && (
          <CopyText text={copyText} hideCopyText={hideCopyText} color={color} />
        )}
      </LabeledLinkEl>
    </LabeledText>
  )
}

export interface LabeledLinkProps {
  label?: string
  color?: string
  href: string
  linkText?: string
  copyText?: string
  disableLink?: boolean
  hideCopy?: boolean
  hideCopyText?: boolean
  useNewTabIcon?: boolean
}

export const LabeledLinkPropTypes = {
  label: PropTypes.string,
  color: PropTypes.string,
  href: PropTypes.string.isRequired,
  linkText: PropTypes.string,
  copyText: PropTypes.string,
  disableLink: PropTypes.bool,
  hideCopy: PropTypes.bool,
  hideCopyText: PropTypes.bool,
  useNewTabIcon: PropTypes.bool
}

LabeledLink.propTypes = LabeledLinkPropTypes
LabeledLink.defaultProps = {
  color: Colors.PURPLE_MEDIUM,
  disableLink: false,
  hideCopy: false,
  hideCopyText: true,
  useNewTabIcon: false
}
