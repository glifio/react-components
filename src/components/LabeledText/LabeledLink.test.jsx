import { render } from '@testing-library/react'
import { LabeledLink } from './LabeledLink'

describe('LabeledLink', () => {
  test('it renders correctly', () => {
    const { container } = render(
      <LabeledLink
        label='Glif Safe'
        href='https://safe.glif.link/'
        linkText='safe.glif.link'
        copyText='https://safe.glif.link/'
      />
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
