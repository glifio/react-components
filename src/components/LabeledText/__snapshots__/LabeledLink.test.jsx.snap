// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabeledLink it renders correctly 1`] = `
.c0 h4 {
  margin: 0;
  color: var(--gray-dark);
}

.c0 p {
  margin: 0;
  line-height: 1.5;
}

.c3 {
  vertical-align: middle;
}

.c2 {
  position: relative;
  cursor: pointer;
}

.c2 svg {
  -webkit-transition: -webkit-transform 0.1s ease-out;
  -webkit-transition: transform 0.1s ease-out;
  transition: transform 0.1s ease-out;
}

.c2 .copy-text {
  position: absolute;
  top: calc(50% - 0.2em);
  left: calc(100%);
  padding-left: 0.5em;
  color: var(--purple-medium);
  opacity: 0;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: opacity 0.2s ease-in;
  transition: opacity 0.2s ease-in;
  pointer-events: none;
  white-space: nowrap;
}

.c2:hover svg {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
}

.c2:hover .copy-text {
  opacity: 1;
  pointer-events: all;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.35em;
  line-height: 1.5;
}

.c1 > a {
  color: var(--purple-medium);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1 > a:hover {
  color: var(--purple-medium);
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1 > a svg {
  -webkit-transition: 0.24s ease-in-out;
  transition: 0.24s ease-in-out;
}

.c1 > a svg:hover {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
  fill: var(--purple-medium);
}

<div
  class="c0"
>
  <h4>
    Glif Safe
  </h4>
  <div
    class="c1"
    color="var(--purple-medium)"
  >
    <a
      href="https://safe.glif.link/"
      rel="noreferrer noopener"
      target="_blank"
    >
      safe.glif.link
    </a>
    <span
      class="c2"
      color="var(--purple-medium)"
      role="button"
    >
      <svg
        class="c3"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="#1A0066"
          height="12"
          width="12"
          x="4"
          y="0"
        />
        <rect
          fill="#E0D7FE"
          height="12"
          width="12"
          x="0"
          y="4"
        />
      </svg>
    </span>
  </div>
</div>
`;
